Add Death Indicator Plugin

DannysPVM opened the pull request:
Add Death Indicator Plugin. On player death, it displays: a hint arrow to death location, an item reclaim timer with death world tooltip, and marks the location of death on the world map.

Edit: Adding a gif of the basic functionality. Note, after this gif was recorded, I did add a tooltip to the world map death marker that says, "Death Location", so that there is no confusion. 

_Basic Functionality of the Death Indicator Plugin_
https://puu.sh/As5XO.gif 

FINAL EDIT: I deleted the repository + branch that I used when creating this PR (Big mistake), so this PR is now closed. I am opening a new PR that will house this plugin. I will also still be referencing this old PR for any recommendation/suggestion that I may not have incorporated yet. Sorry for any trouble I may have caused.

DannysPVM commented:
Current version does everything requested in issue #2864 except for save information on client crash. I am working on saving the info on client crash, but in the meantime this current version is alright to be implemented, of course though, it's not up to me d:

Reasel commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event)
+	{
+		if (event.getGroup().equals("deathindicator") && !config.showDeathHintArrow())
```
Maybe I am being picky, but it seems like you should only check for deathindicator changes once. Then inside that if check each config.

DannysPVM replied:
So like a nested if, where the outter if (event.getGroup().equals("deathindicator")) then the inner if's being if (!config.BLAHBLAHBLAH)? I could get behind that.

Reasel replied:
Yes

DannysPVM replied:
Will change soon, working on getting the death information saved via hidden config items and upon login, if the player has died and not returned to their items, showing the death information. The current version I've PR'd here does this upon logout/hopping, but on client crash the player is sol atm. If I can't get the saving information and pulling it back up on relaunch to work properly, I'll just update the previous version to meet your request and push that. :)

Reasel replied:
If you need an example of hidden config I think the farming plugin utilizes this.

DannysPVM replied:
Fixed, thanks for the tip :)

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorConfig.java:
```diff
@@ -0,0 +1,46 @@
+@ConfigGroup(
+		keyName = "deathindicator",
+		name = "Death Indicator",
+		description = "Configuration for the death indicator plugin"
```
Indent should be fixed, these should only be 1 tab

DannysPVM replied:
noted on the indents, thanks, will fix.

DannysPVM replied:
fixed indents, thanks for the review :)

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorConfig.java:
```diff
@@ -0,0 +1,46 @@
+			position = 1,
+			keyName = "deathhintarrow",
+			name = "Death Hint Arrow",
+			description = "Configures whether or not to show a hint arrow to death location"
```
Same for these, these should only be 2 tabs (Same with the ones below)

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+		if (!playerHasRespawned)
+		{
+			return;
+		}
```
This should just be moved to the statement above

DannysPVM replied:
I'm not exactly sure why, but when I do combine these two statements the entire plugin breaks. I think it's because they fire on game tick events and the death occurs 5 seconds before the respawn occurs so if they are together I miss the opportunity to move forward and document the death.

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+		if (currentWorld != deathWorld)
+		{
+			return;
+		}
```
Is this supposed to just return or is it in progress?

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+import net.runelite.client.ui.overlay.worldmap.WorldMapPointManager;
+
+@PluginDescriptor(
+		name = "Death Indicator"
```
Indent here as well

DannysPVM replied:
Fixed, thanks!

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+	private boolean worldMapPointSet = false;
+	private boolean hasDied = false;
+	private Timer deathTimer = null;
+	private static int currentWorld = -1;
```
Double check Java standards, I believe static should be before noon static variables

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+	private DeathIndicatorConfig config;
+
+	@Inject
+	private DeathIndicatorPlugin plugin;
```
Use `this` instead of injecting a self reference

raqes replied:
reference to the DeathIndicatorPlugin by using the aforementioned `this` for example `deathTimer = new Timer(60, ChronoUnit.MINUTES, BONES, this);`

MESLewis replied:
Remove this line

DannysPVM replied:
Oh okay, I got it, thanks guys I had no idea lol

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+		currentLocation = player.getWorldLocation();
+
+		if ((currentLocation.getX() == deathLocation.getX())
+				&& (currentLocation.getY() == deathLocation.getY())
```
This indent can also be moved back a tab

DannysPVM replied:
Fixed, thanks man!

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+		}
+	}
+
+	private void addMapPoint(WorldPoint point)
```
This can be in lined to be cleaner. Always remove the point before adding it and if it isn't already there it will remove nothing and be fine.
This also eliminates the Boolean

DannysPVM replied:
Done, Thank you!

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathWorldMapPoint.java:
```diff
@@ -0,0 +1,48 @@
+package net.runelite.client.plugins.deathindicator;
```
License header

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,267 @@
+package net.runelite.client.plugins.deathindicator;
```
Header

MESLewis commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorConfig.java:
```diff
@@ -0,0 +1,46 @@
+package net.runelite.client.plugins.deathindicator;
```
Header

DannysPVM commented:
Thank you all for the feedback, I am just about finished making changes to support if the client closes/crashes saving the death information and showing it upon client relaunch. Before I push those changes I'll make sure to fix each and every thing noted here!

DannysPVM commented:
Implemented the recommended changes and added support for saving death location, world, and time left to reclaim upon client close and/or crash. Closing this PR, and opening a new PR.

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorConfig.java:
```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2017, Adam <Adam@sigterm.info>
```
Is this copyright correct? New files only need your own name unless you mean to credit it to the RuneLite project specifically.

DannysPVM replied:
It's not, I'll go ahead and remove adam's name from the copyright :P (Copypaste strikes again).

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,355 @@
+			config.hasRespawned(true);
+		}
+
+		if ((config.hasDied()) && (!config.timeOfDeath().equals("")))
```
To move one statement out of the heavily nested if statements, you could do:
```java
if (!config.hasDied() && config.timeofDeath().equals(""))
{
    return;
}
```

DannysPVM replied:
I like that idea, implemented this change to remove one of those nested loops, it looks much better now. Thanks!

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,355 @@
+			return;
+		}
+
+		if (client.getWorld() != config.deathWorld())
```
This should be moved to the first if statement

DannysPVM replied:
You may have to just show me what you want it to look like, I can't get the plugin to work if I have the has died and has respawned on the same line or nested inside each other, and the same for the current world vs death world. Like I commented earlier, I think its because the events occur on different game ticks and if I miss the chance to continue down then the death doesn't get reccorded (it gets marked then instantly removed), but I may also just be dumb and not be doing what you want.

ypperlig replied:
Since you're just returning until all statements are true:
```java
if (!config.hasDied() || !config.hasRespawned() || client.getWorld() != config.deathWorld())
{
    return;
}
```

DannysPVM replied:
Works perfectly man, thanks! I was stuck on the idea of using && rather than ||, my brain gets locked into an idea and I go blind to the obvious. I appreciate the clarification on that.

ypperlig commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,355 @@
+			return;
+		}
+
+		if (!config.hasRespawned())
```
This should be moved to the first if statement

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,353 @@
+
+		Actor actor = event.getActor();
+
+		if (actor instanceof NPC || actor != client.getLocalPlayer())
```
this can just be `actor != client.getLocalPlayer()`

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,353 @@
+	public static final BufferedImage WORLDMAP_ARROW;
+	public static final BufferedImage BONES;
+
+	private WorldPoint deathLocation;
```
It doesn't look like this has a purpose

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,353 @@
+	@Subscribe
+	public void onChatMessage(ChatMessage event)
+	{
+		if (event.getMessage().equals("Oh dear, you are dead!"))
```
You need to check chat message type here

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,353 @@
+
+			if (deathTimer != null)
+			{
+				infoBoxManager.removeInfoBox(deathTimer);
```
I would null out deathTimer everywhere it is removed

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,353 @@
+				infoBoxManager.removeInfoBox(deathTimer);
+			}
+
+			if (config.worldMapPointSet())
```
Theres no point tracking this, just remove it if it exists.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,353 @@
+			return;
+		}
+
+		if (client.getWorld() == config.deathWorld())
```
There should be a check for LOGGED_IN here

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,353 @@
+			config.worldMapPointSet(false);
+		}
+
+		if (event.getGameState() == GameState.LOGIN_SCREEN || event.getGameState() == GameState.CONNECTION_LOST)
```
As far as I can tell this is unnecessary

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorConfig.java:
```diff
@@ -0,0 +1,235 @@
+		description = "",
+		hidden = true
+	)
+	default String timeOfDeath()
```
Config supports Instant

LeviSchuck replied:
ConfigManager says otherwise

SoyChai replied:
Yeah, I'm not sure what Adam meant here either.

DannysPVM replied:
I believe I've fixed everything else mentioned at this point. I'm going to go ahead and proofread my code again then submit it with storing the death time as a string. If I need to change that later, I'll try to figure it out.

DannysPVM replied:
This is the last unresolved comment @Adam- . I'm sorry that I didn't clean up the commits to make it look nicer, this is my first PR ever, and I screwed up my branch and can't fix it. I tried for about an hour and fifteen minutes to fix my branch so I could squash some pointless commits but it just wasn't happening. :/ Let me know If I need to do anything else.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+			synchronized (ImageIO.class)
+			{
+				WORLDMAP_ARROW = ImageIO.read(DeathIndicatorPlugin.class.getResourceAsStream("worldmap_arrow.png"));
+				BONES = ImageIO.read(DeathIndicatorPlugin.class.getResourceAsStream("bones.png"));
```
You can get this image from ItemManager

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+		{
+			synchronized (ImageIO.class)
+			{
+				WORLDMAP_ARROW = ImageIO.read(DeathIndicatorPlugin.class.getResourceAsStream("worldmap_arrow.png"));
```
I would probably just move the arrow from clue scroll plugin and reuse that image, because it is same.

DannysPVM replied:
Fixed that to get the arrow from ClueScrollPlugin, but having some trouble rendering the bones from ItemManager.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+	}
+
+	@Override
+	protected void shutDown() throws Exception
```
what about startup? you should try to create the marker on startup too.

DannysPVM replied:
I was personally using, a ChatMessage eventm if the message is SERVER type and is "Welcome to RuneScape." as my startup, but if I need to, I can change that to startUp().

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+			{
+				if (config.showDeathHintArrow())
+				{
+					client.setHintArrow(new WorldPoint(config.deathLocationX(), config.deathLocationY(), config.deathLocationPlane()));
```
This arrow will get deleted when you f.e get bounty hunter target or do quests/clues etc (same with the onActorDeath)

DannysPVM replied:
I'm not sure exactly what to do to fix that issue, can you elaborate a bit more? Thanks :)

DannysPVM replied:
So @deathbeam , should I just do something like:

if (client.hasHintArrow())
{
return;
}
else
{
client.setHintArrow(...);
}

That way if a different arrow already exists I don't override it/clear it?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+				&& (config.deathLocationY() == -1)
+				&& (config.deathLocationPlane() == -1)
+				&& (!config.hasDied())
+				&& (config.hasRespawned())
```
Why you need hasRespawned and hasDied? All you need is null-check timeOfDeath

DannysPVM replied:
I was using hasDied as a check (for efficency purposes) to make sure the player has at least died before doing something, the hasRespawned comes into play for resetting the markers. If I don't have the hasRespawned boolean, the marker is set and then instantly cleared.

Kamielvf commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+
+		if (config.showDeathInfoBox())
+		{
+			config.timeOfDeath(Instant.now());
```
I would always save this, in case the user toggles on the info box you would need this data.

Kamielvf commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+			config.timeOfDeath(Instant.now());
+			deathTimer = new Timer(60, ChronoUnit.MINUTES, BONES, this);
+			deathTimer.setTooltip("Died on world: " + config.deathWorld());
+			infoBoxManager.addInfoBox(deathTimer);
```
Make a method something like `addDeathTimer` because this code is ran from multiple places.

DannysPVM replied:
Will do, thanks Kamiel!

Kamielvf commented on runelite-client/src/main/java/net/runelite/client/plugins/deathindicator/DeathIndicatorPlugin.java:
```diff
@@ -0,0 +1,311 @@
+					}
+					deathTimer = new Timer(Duration.ofHours(1).minus(Duration.between(config.timeOfDeath(), Instant.now())).getSeconds(), ChronoUnit.SECONDS, BONES, this);
+					deathTimer.setTooltip("Died on world: " + config.deathWorld());
+					infoBoxManager.addInfoBox(deathTimer);
```
see `addDeathTimer ` comment above
